### PR TITLE
Extra value types for xframe

### PR DESCRIPTION
**Pull request by linkRace**
Adds options to have xframe header set on blocklist/allowlist format or by custom defined functions.

Tests added for all cases.

### CODE REVIEW

**Comment by linkRace** [comment]
@grawk looking to publish a `1.7.0` with this and #143 , let me know what you think.

File: README.md [hunk 1]
```diff
@@ -62,8 +62,10 @@ __Please note that you must use [express-session](https://github.com/expressjs/s
 * `angular` Boolean - Optional. Shorthand setting to set `lusca` up to use the default settings for CSRF validation according to the [AngularJS docs]. Can be used with `cookie.options`.
 * `blocklist` Array or String - Optional. Allows defining a set of routes that will not have csrf protection.  All others will.
 * `allowlist` Array or String - Optional. Allows defining a set of routes that will have csrf protection.  All others will not.
+* `csrfFunction` Function - Returns `true` if route should require a token or `false` if route should not require a token.
```
Comment (maxmil7) [inline README.md:65]: we can mark it as Optional

File: lib/csrf.js [hunk 2]
```diff
@@ -92,17 +92,16 @@ module.exports = function (options) {
         var shouldBypass = false;
 
         if (blocklist) {
-            blocklist.some(function (exclusion) {
-                shouldBypass = req.path.indexOf(exclusion) === 0;
-                return shouldBypass;
-            });
+            shouldBypass = (blocklist.indexOf(req.path) !== -1);
```
Comment (maxmil7) [inline lib/csrf.js:95]: this seems to change the functionality. For eg. earlier `/foo` in blocklist would return true for `/foo/` but this won't

File: lib/csrf.js [hunk 3]
```diff
@@ -92,17 +92,16 @@ module.exports = function (options) {
...
         }
 
         if (allowlist) {
-            allowlist.some(function (inclusion) {
-                shouldBypass = req.path.indexOf(inclusion) !== 0;
-                return shouldBypass;
-            });
+            shouldBypass = (allowlist.indexOf(req.path) === -1);
+        }
```
Comment (maxmil7) [inline lib/csrf.js:100]: same as previous comment

File: lib/xframes.js [hunk 4]
```diff
@@ -6,9 +6,54 @@
...
+            }
+
+            if (allowlist) {
+                shouldBypass = (allowlist.indexOf(req.path) === -1);
+            }
+
+            xframeFunction = options.xframeFunction;
+            if (typeof xframeFunction === 'function') {
+                shouldBypass = false;
+                headerValue = xframeFunction(req);
```
Comment (maxmil7) [inline lib/xframes.js:46]: we should set `shouldBypass` to true, if function returns false

File: lib/xframes.js [hunk 5]
```diff
@@ -6,9 +6,54 @@
...
+
+            if (typeof blocklist === 'string') {
+                blocklist = [blocklist];
+            } else if (!Array.isArray(blocklist)) {
+                // Don't allow non string or array blocklist
+                blocklist = null;
+            }
+
+            if (blocklist) {
+                shouldBypass = (blocklist.indexOf(req.path) !== -1);
```
Comment (maxmil7) [inline lib/xframes.js:36]: same issue as one in csrf blocklist check

File: test/csrf.js [hunk 6]
```diff
@@ -45,11 +45,11 @@ describe('CSRF', function () {
     it('should not require token on post to blocklist', function (done) {
         var app = mock({
             csrf: {
-                blocklist: ['/blocklist1', '/blocklist2']
+                blocklist: ['/blocklist1', '/blocklist2', '/']
```
Comment (maxmil7) [inline test/csrf.js:48]: are we using blocklist1 anywhere?

File: test/csrf.js [hunk 7]
```diff
@@ -81,7 +81,7 @@ describe('CSRF', function () {
     it('should only require token on post to allowlist', function (done) {
         var app = mock({
             csrf: {
-                allowlist: ['/allowlist1', '/allowlist2']
+                allowlist: ['/', '/allowlist1', '/allowlist2']
```
Comment (maxmil7) [inline test/csrf.js:84]: please add a test for trailing `/` as well

File: test/csrf.js [hunk 8]
```diff
@@ -110,10 +114,67 @@ describe('CSRF', function () {
...
+            .expect(403)
             .end(function (err, res) {
                 done(err);
             });
     });
+
+    it('should only require token on post to whitelist 2', function (done) {
+        var app = mock({
+            csrf: {
+                whitelist: ['/api/buttons', '/api/data']
```
Comment (maxmil7) [inline test/csrf.js:130]: is `/api/data` being used anywhere?

File: lib/csrf.js (hunk 2 above)
Comment (linkRace) [reply lib/csrf.js:95]: ah, interesting. problem we were running into is that `/` in the blocklist would put all routes into the blocklist.

File: README.md (hunk 1 above)
Comment (linkRace) [reply README.md:65]: will change

File: test/xframe.js [hunk 9]
```diff
@@ -44,4 +44,120 @@ describe('XFRAME', function () {
...
+        app.get('/allow', function (req, res) {
+            res.status(200).end();
+        });
+
+        request(app)
+            .get('/allow')
+            .expect(200)
+            .end(function (err, res) {
+                var isHeaderPresent = res.header['x-frame-options'] !== undefined;
+                assert(isHeaderPresent);
```
Comment (maxmil7) [inline test/xframe.js:78]: any reason not to use the expect syntax similar to line 57?

File: test/xframe.js [hunk 10]
```diff
@@ -44,4 +44,120 @@ describe('XFRAME', function () {
...
+            .get('/')
+            .expect(200)
+            .end(function (err, res) {
+                var isHeaderPresent = res.header['x-frame-options'] !== undefined;
+                assert(!isHeaderPresent);
+                done();
+            });
+    });
+
+    it('header (sameorigin) not on blocklist', function (done) {
```
Comment (maxmil7) [inline test/xframe.js:133]: this use case seems to be covered in previous test case

File: test/xframe.js [hunk 11]
```diff
@@ -44,4 +44,120 @@ describe('XFRAME', function () {
...
+        });
+
+        request(app)
+            .get('/')
+            .expect('X-FRAME-OPTIONS', 'SAMEORIGIN')
+            .expect(200, done);
+    });
+
+    it('header function', function (done) {
+        var config = { xframe: {xframeFunction: function (req) {
```
Comment (maxmil7) [inline test/xframe.js:148]: can we add a test for when function returns `false`

File: test/xframe.js (hunk 10 above)
Comment (maxmil7) [reply test/xframe.js:133]: might be good to add a scenario where `allowlist` or `blocklist` is a string

File: test/xframe.js (hunk 9 above)
Comment (linkRace) [reply test/xframe.js:78]: good point, will change

**Comment by linkRace** [comment]
Believe I've addressed all the comments outside of the trailing / issue.
I could add in logic checking for a route name and/or a trailing /. Wouldn't want to break any existing functionality with a minor version bump, but I'm not sure that's the intended behavior. Let me know your thoughts.

File: lib/csrf.js (hunk 2 above)
Comment (maxmil7) [reply lib/csrf.js:95]: I think we can add an explicit check for `/` in config. For eg:
```
blocklist.some(function (exclusion) {
  shouldBypass = exclusion.length === 1 ? req.path === exclusion : req.path.indexOf(exclusion) === 0 :
  return shouldBypass;
})
     
```

Comment (maxmil7) [reply lib/csrf.js:95]: @grawk any thoughts?

Comment (linkRace) [reply lib/csrf.js:95]: how about a case where the blocklist defines `/blocklist` and the route `/blocklist/extra/` is passed? 

should `/blocklist/extra/` be on the blocklist or only if `/blocklist/*` is specified?

**Comment by maxmil7** [comment]
Change partially implemented in #145 
Closing this PR for the time being.

**Pull request closed without merging**